### PR TITLE
feat(cli): Support `nested_type` in generator

### PR DESCRIPTION
**Pull request by ansgarm**
Support for tfproto v6

### Open Tasks
- [ ] Write Test(s) for various `nested_type` configurations
- [ ] Write test for minItems & maxItems behaviour!
- [ ] Address all `FIXME` comments

### CODE REVIEW

File: packages/cdktf-cli/lib/get/generator/emitter/struct-emitter.ts [hunk 1]
```diff
@@ -65,6 +65,10 @@ export class StructEmitter {
       );
     }
     this.code.closeBlock();
+
+    if (!(struct instanceof ConfigStruct)) {
+      this.emitToTerraformFuction(struct);
+    }
   }
```
Comment (jsteinich) [inline packages/cdktf-cli/lib/get/generator/emitter/struct-emitter.ts:72]: I believe classes are only emitted for computed types.

File: packages/cdktf-cli/lib/get/generator/models/attribute-model.ts [hunk 2]
```diff
@@ -75,6 +75,8 @@ export class AttributeModel {
     if (this._name === "equals") return "equalTo";
     // `node` is already used by the Constructs base class
     if (this._name === "node") return "nodeAttribute";
+    // FIXME: this is only temporary and will not work when this attribute is actually used.
+    if (this._name === "provider") return "providerAttribute";
```
Comment (jsteinich) [inline packages/cdktf-cli/lib/get/generator/models/attribute-model.ts:79]: Seems unrelated. Did this come up?

File: packages/cdktf-cli/lib/get/generator/models/attribute-type-model.ts [hunk 3]
```diff
@@ -30,6 +30,7 @@ export class AttributeTypeModel {
   public isMap: boolean;
   public struct?: Struct;
   public level?: number;
+  // FIXME: do we need an isNested here? maybe for creating different outcomes for the TF JSON?
```
Comment (jsteinich) [inline packages/cdktf-cli/lib/get/generator/models/attribute-type-model.ts:33]: Other than the actual proto definition file, I'm having trouble finding documentation on the changes to know if anything is different.

File: packages/cdktf-cli/lib/get/generator/provider-schema.ts [hunk 4]
```diff
@@ -24,15 +28,68 @@ export interface Schema {
...
 
-export interface Attribute {
-  type: AttributeType;
+type AttributeNestedTypeNesting =
+  | "invalid"
+  | "single"
+  | "list"
+  | "set"
+  | "map"
+  | "group";
```
Comment (jsteinich) [inline packages/cdktf-cli/lib/get/generator/provider-schema.ts:37]: `group` isn't valid for nested types, just regular types. I'm also not sure what it does is that case.

File: packages/cdktf-cli/lib/get/generator/provider-schema.ts [hunk 5]
```diff
@@ -43,7 +100,7 @@ export type AttributeType =
   | ["list", AttributeType]
   | ["object", { [attribute: string]: AttributeType }];
 
-export type BlockType =
+export type BlockType =  // FIXME: use such a type for NestedType Attributes as well?
```
Comment (jsteinich) [inline packages/cdktf-cli/lib/get/generator/provider-schema.ts:103]: Might not hurt, but seems to be over specifying compared to the proto definition.

File: packages/cdktf-cli/lib/get/generator/resource-parser.ts [hunk 6]
```diff
@@ -187,7 +190,48 @@ class Parser {
...
+        case "set": {
+          isList = true; // FIXME: check required / optional based on min_items
+          break;
+        }
+        case "map":
+          isMap = true;
+          break;
+        case "single":
+      }
+      const struct = this.addAnonymousStruct(scope, attributeType.attributes); // FIXME: don't make anonymous?
```
Comment (jsteinich) [inline packages/cdktf-cli/lib/get/generator/resource-parser.ts:219]: Depends on if a block type exists elsewhere.

**Review by jsteinich — COMMENTED** [review]
Is there another source of documentation besides the proto definition file?

File: packages/cdktf-cli/lib/get/generator/models/attribute-model.ts (hunk 2 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/get/generator/models/attribute-model.ts:79]: The latest commit is a "wip" commit 😁 So yeah, that'll go again, but currently I need this for the provider I'm testing with (some private internal one which is also currently wip, so that's why it had this attribute I guess).

File: packages/cdktf-cli/lib/get/generator/models/attribute-type-model.ts (hunk 3 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/get/generator/models/attribute-type-model.ts:33]: I have not found anything yet. There is an internal RFC describing the protocol v6 (which seems to be only adding the nested type) but that is a bit broader.

File: packages/cdktf-cli/lib/get/generator/provider-schema.ts (hunk 4 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/get/generator/provider-schema.ts:37]: You're right – I took that from the internal RFC but that does not seem to have made the cut to the implementation / or didn't make sense to implement 👍

File: packages/cdktf-cli/lib/get/generator/provider-schema.ts (hunk 5 above)
Comment (ansgarm) [reply packages/cdktf-cli/lib/get/generator/provider-schema.ts:103]: yeah, right.

**Comment by skorfmann** [comment]
@ansgarm is there a path how we can merge this soonish, or should we close this for now and revisit when the need surfaces again?

**Comment by jsteinich** [comment]
> @ansgarm is there a path how we can merge this soonish, or should we close this for now and revisit when the need surfaces again?

I have this on my near-term todo list, but I was planning to just make a new PR since there are a few other pieces mixed in here.

**Comment by skorfmann** [comment]
Closing as suggested by @ansgarm
